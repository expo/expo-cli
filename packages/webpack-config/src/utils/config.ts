function isObject(val: any): boolean {
  if (val === null) {
    return false;
  }
  return typeof val === 'function' || typeof val === 'object';
}

/**
 * Given a config option that could evalutate to true, config, or null; return a config.
 * e.g.
 * `polyfill: true` returns the `config`
 * `polyfill: {}` returns `{}`
 *
 * @category utils
 */
export function enableWithPropertyOrConfig(
  prop: any,
  config: boolean | { [key: string]: any },
  merge: boolean = false
): any {
  // Value is truthy.
  if (prop) {
    if (isObject(prop)) {
      if (merge) {
        if (config == null || typeof config !== 'object') {
          throw new Error('enableWithPropertyOrConfig cannot merge config: ' + config);
        }
        return {
          ...config,
          ...prop,
        };
      }

      // Return property
      return prop;
    }

    // Value is truthy but not a replacement config, thus return the default config.
    return config;
  }
  // Return falsey.
  return prop;
}

/**
 * Used for features that are enabled by default unless specified otherwise.
 *
 * @category utils
 */
export function overrideWithPropertyOrConfig(
  prop: any,
  config: boolean | { [key: string]: any },
  merge: boolean = false
): any {
  if (prop === undefined) {
    return config;
  }
  return enableWithPropertyOrConfig(prop, config, merge);
}
