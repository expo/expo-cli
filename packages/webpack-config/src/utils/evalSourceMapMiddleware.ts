/**
 * Copyright (c) 2022 Expo, Inc.
 * Copyright (c) 2015-present, Facebook, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * Based on https://github.com/facebook/create-react-app/blob/a422bf2/packages/react-dev-utils/evalSourceMapMiddleware.js
 * But with Node LTS support.
 */

import type express from 'express';
import type webpack from 'webpack';
import type WebpackDevServer from 'webpack-dev-server';

function base64SourceMap(source: any) {
  const base64 = Buffer.from(JSON.stringify(source.map()), 'utf8').toString('base64');
  return `data:application/json;charset=utf-8;base64,${base64}`;
}

function getSourceById(server: WebpackDevServer & any, id: string) {
  const module = Array.from(server._stats.compilation.modules).find(
    m => server._stats.compilation.chunkGraph.getModuleId(m) === id
  ) as webpack.Module[];

  // @ts-ignore
  return module.originalSource();
}

/*
 * Middleware responsible for retrieving a generated source
 * Receives a webpack internal url: "webpack-internal:///<module-id>"
 * Returns a generated source: "<source-text><sourceMappingURL><sourceURL>"
 *
 * Based on EvalSourceMapDevToolModuleTemplatePlugin.js
 */
export function createEvalSourceMapMiddleware(server: WebpackDevServer) {
  return function handleWebpackInternalMiddleware(
    req: express.Request,
    res: express.Response,
    next: express.NextFunction
  ) {
    if (req.url.startsWith('/__get-internal-source')) {
      const fileName = req.query.fileName;
      if (typeof fileName !== 'string') {
        return next();
      }
      const id = fileName?.match(/webpack-internal:\/\/\/(.+)/)?.[1];
      // @ts-ignore: untyped
      if (!id || !server._stats) {
        return next();
      }

      const source = getSourceById(server, id);
      const sourceMapURL = `//# sourceMappingURL=${base64SourceMap(source)}`;
      const sourceURL = `//# sourceURL=webpack-internal:///${module.id}`;
      res.end(`${source.source()}\n${sourceMapURL}\n${sourceURL}`);
    } else {
      return next();
    }
  };
}
