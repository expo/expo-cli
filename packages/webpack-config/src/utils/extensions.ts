export function getModuleFileExtensionsWithoutDotPrefix(...platforms: string[]): string[] {
  let fileExtensions = [];

  // Support both TypeScript and JavaScript
  for (const extension of ['ts', 'tsx', 'js', 'jsx']) {
    // Ensure order is correct: [platformA.js, platformB.js, js]
    for (const platform of [...platforms, '']) {
      fileExtensions.push([platform, extension].filter(Boolean).join('.'));
    }
  }
  // Always add this last
  fileExtensions.push('json');

  return fileExtensions;
}

export function getModuleFileExtensions(...platforms: string[]): string[] {
  // Webpack requires a `.` before each value
  return getModuleFileExtensionsWithoutDotPrefix(...platforms).map(value => `.${value}`);
}
