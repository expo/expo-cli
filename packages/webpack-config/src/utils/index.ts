export * from './search';
export { default as normalizePaths } from './normalizePaths';
export * from './config';
