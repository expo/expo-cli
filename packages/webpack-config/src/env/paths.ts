/* eslint-env node */
import { ExpoConfig, getConfig, getWebOutputPath } from '@expo/config';
import {
  ensureSlash,
  getAbsolutePathWithProjectRoot,
  getEntryPoint,
  getModulesPath,
  getPossibleProjectRoot,
} from '@expo/config/paths';
import fs from 'fs';
import path from 'path';
import url from 'url';

import { Environment, FilePaths, Mode } from '../types';
import getMode from './getMode';
import { getConfigMode } from './getConfigMode';

function parsePaths(projectRoot: string, mode: Mode, nativeAppManifest?: ExpoConfig): FilePaths {
  const inputProjectRoot = projectRoot || getPossibleProjectRoot();

  function absolute(...pathComponents: string[]): string {
    return getAbsolutePathWithProjectRoot(inputProjectRoot, ...pathComponents);
  }

  const packageJsonPath = absolute('package.json');
  const modulesPath = getModulesPath(inputProjectRoot);
  const productionPath = absolute(getWebOutputPath(nativeAppManifest));

  function templatePath(filename: string = ''): string {
    const overridePath = absolute('web', filename);
    if (fs.existsSync(overridePath)) {
      return overridePath;
    }
    return path.join(__dirname, '../../web-default', filename);
  }

  function getProductionPath(...props: string[]): string {
    return path.resolve(productionPath, ...props);
  }

  function getIncludeModule(...props: string[]): string {
    return path.resolve(modulesPath, ...props);
  }

  let appMain: string | null = null;
  try {
    appMain = getEntryPoint(inputProjectRoot, ['./index', './src/index'], ['web']);
  } catch (_) {
    // ignore the error
  }

  return {
    absolute,
    includeModule: getIncludeModule,
    packageJson: packageJsonPath,
    root: path.resolve(inputProjectRoot),
    appMain,
    modules: modulesPath,
    servedPath: getServedPath(inputProjectRoot, mode),
    template: {
      get: templatePath,
      folder: templatePath(),
      indexHtml: templatePath('index.html'),
      manifest: templatePath('manifest.json'),
      serveJson: templatePath('serve.json'),
      favicon: templatePath('favicon.ico'),
      serviceWorker: templatePath('expo-service-worker.js'),
      registerServiceWorker: templatePath('register-service-worker.js'),
    },
    production: {
      get: getProductionPath,
      folder: getProductionPath(),
      indexHtml: getProductionPath('index.html'),
      manifest: getProductionPath('manifest.json'),
      serveJson: getProductionPath('serve.json'),
      favicon: getProductionPath('favicon.ico'),
      serviceWorker: getProductionPath('expo-service-worker.js'),
      registerServiceWorker: getProductionPath('register-service-worker.js'),
    },
  };
}

/**
 * Sync method for getting default paths used throughout the Webpack config.
 * This is useful for Next.js which doesn't support async Webpack configs.
 *
 * @param projectRoot
 * @category env
 */
export function getPaths(projectRoot: string, mode?: string): FilePaths {
  const configMode = getConfigMode(getMode({ mode }));
  const { exp } = getConfig(projectRoot, {
    skipSDKVersionRequirement: true,
    mode: configMode,
  });
  return parsePaths(projectRoot, configMode, exp);
}

/**
 * Async method for getting default paths used throughout the Webpack config.
 *
 * @param projectRoot
 * @category env
 */
export async function getPathsAsync(projectRoot: string, mode?: string): Promise<FilePaths> {
  const configMode = getConfigMode(getMode({ mode }));

  let exp;
  try {
    exp = getConfig(projectRoot, { skipSDKVersionRequirement: true, mode: configMode }).exp;
  } catch (error) {}
  return parsePaths(projectRoot, configMode, exp);
}

/**
 * Get paths dictating where the app is served regardless of the current Webpack mode.
 *
 * @param projectRoot
 * @category env
 */
export function getServedPath(projectRoot: string, mode?: string): string {
  const configMode = getConfigMode(getMode({ mode }));

  const { pkg } = getConfig(projectRoot, {
    skipSDKVersionRequirement: true,
    mode: configMode,
  });
  const envPublicUrl = process.env.WEB_PUBLIC_URL;

  // We use `WEB_PUBLIC_URL` environment variable or "homepage" field to infer
  // "public path" at which the app is served.
  // Webpack needs to know it to put the right <script> hrefs into HTML even in
  // single-page apps that may serve index.html for nested URLs like /todos/42.
  // We can't use a relative path in HTML because we don't want to load something
  // like /todos/42/static/js/bundle.7289d.js. We have to know the root.
  const publicUrl = envPublicUrl || pkg.homepage;
  const servedUrl = envPublicUrl || (publicUrl ? url.parse(publicUrl).pathname : '/');
  return ensureSlash(servedUrl!, true);
}

/**
 * Get paths dictating where the app is served. In development mode this returns default values.
 *
 * @param env
 * @category env
 */
export function getPublicPaths(
  env: Pick<Environment, 'mode' | 'projectRoot'>
): {
  /**
   * Webpack uses `publicPath` to determine where the app is being served from.
   * It requires a trailing slash, or the file assets will get an incorrect path.
   * In development, we always serve from the root. This makes config easier.
   */
  publicPath: string;

  /**
   * `publicUrl` is just like `publicPath`, but we will provide it to our app
   * as %WEB_PUBLIC_URL% in `index.html` and `process.env.WEB_PUBLIC_URL` in JavaScript.
   * Omit trailing slash as %WEB_PUBLIC_URL%/xyz looks better than %WEB_PUBLIC_URL%xyz.
   */
  publicUrl: string;
} {
  const parsedMode = getMode(env);
  if (parsedMode === 'production') {
    const publicPath = getServedPath(env.projectRoot, parsedMode);
    return {
      publicPath,
      publicUrl: publicPath.slice(0, -1),
    };
  }

  return { publicUrl: '', publicPath: '/' };
}

/**
 * Get the output folder path. Defaults to `web-build`.
 *
 * @param projectRoot
 * @category env
 */
export function getProductionPath(projectRoot: string, mode?: string): string {
  const configMode = getConfigMode(getMode({ mode }));

  const { exp } = getConfig(projectRoot, {
    skipSDKVersionRequirement: true,
    mode: configMode,
  });
  return getAbsolutePathWithProjectRoot(projectRoot, getWebOutputPath(exp));
}

/**
 * Get an absolute path relative to the project root while accounting for remote paths (`https://`).
 *
 * @param projectRoot
 * @category env
 */
export function getAbsolute(projectRoot: string, ...pathComponents: string[]): string {
  const inputProjectRoot = projectRoot || getPossibleProjectRoot();
  return getAbsolutePathWithProjectRoot(inputProjectRoot, ...pathComponents);
}
