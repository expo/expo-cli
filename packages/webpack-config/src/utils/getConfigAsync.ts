import { ExpoConfig, getConfigForPWA } from '@expo/config';

import { Environment } from '../types';
import getPathsAsync from './getPathsAsync';

async function getConfigAsync(env: Environment): Promise<ExpoConfig> {
  if (env.config) {
    return env.config;
  }
  const locations = await getPathsAsync(env);
  // Fill all config values with PWA defaults
  return getConfigForPWA(env.projectRoot, locations.absolute, {
    templateIcon: locations.template.get('icon.png'),
  });
}

export default getConfigAsync;
