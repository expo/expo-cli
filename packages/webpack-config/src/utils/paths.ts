import { ExpoConfig, getWebOutputPath, readConfigJson, readConfigJsonAsync } from '@expo/config';
import findWorkspaceRoot from 'find-yarn-workspace-root';
import fs from 'fs';
import path from 'path';
import url from 'url';

import { Environment, FilePaths } from '../types';
import getMode from './getMode';

const possibleMainFiles = [
  'index.web.ts',
  'index.ts',
  'index.web.tsx',
  'index.tsx',
  'src/index.web.ts',
  'src/index.ts',
  'src/index.web.tsx',
  'src/index.tsx',
  'index.web.js',
  'index.js',
  'index.web.jsx',
  'index.jsx',
  'src/index.web.js',
  'src/index.js',
  'src/index.web.jsx',
  'src/index.jsx',
  'node_modules/expo/AppEntry.js',
];

function ensureSlash(inputPath: string, needsSlash: boolean): string {
  const hasSlash = inputPath.endsWith('/');
  if (hasSlash && !needsSlash) {
    return inputPath.substr(0, inputPath.length - 1);
  } else if (!hasSlash && needsSlash) {
    return `${inputPath}/`;
  } else {
    return inputPath;
  }
}

export function getAbsolutePathWithProjectRoot(
  projectRoot: string,
  ...pathComponents: string[]
): string {
  // Simple check if we are dealing with an URL
  if (pathComponents && pathComponents.length === 1 && pathComponents[0].startsWith('http')) {
    return pathComponents[0];
  }

  return path.resolve(projectRoot, ...pathComponents);
}

export function getPossibleProjectRoot(): string {
  return fs.realpathSync(process.cwd());
}

export function getModulesPath(projectRoot: string): string {
  const workspaceRoot = findWorkspaceRoot(path.resolve(projectRoot)); // Absolute path or null
  if (workspaceRoot) {
    return path.resolve(workspaceRoot, 'node_modules');
  }

  return path.resolve(projectRoot, 'node_modules');
}

function parsePaths(projectRoot: string, nativeAppManifest?: ExpoConfig): FilePaths {
  const inputProjectRoot = projectRoot || getPossibleProjectRoot();

  function absolute(...pathComponents: string[]): string {
    return getAbsolutePathWithProjectRoot(inputProjectRoot, ...pathComponents);
  }

  const packageJsonPath = absolute('package.json');
  const modulesPath = getModulesPath(inputProjectRoot);
  const productionPath = absolute(getWebOutputPath(nativeAppManifest));

  function templatePath(filename: string = ''): string {
    const overridePath = absolute('web', filename);
    if (fs.existsSync(overridePath)) {
      return overridePath;
    }
    return path.join(__dirname, '../../web-default', filename);
  }

  function getProductionPath(...props: string[]): string {
    return path.resolve(productionPath, ...props);
  }

  function getIncludeModule(...props: string[]): string {
    return path.resolve(modulesPath, ...props);
  }

  return {
    absolute,
    includeModule: getIncludeModule,
    packageJson: packageJsonPath,
    root: path.resolve(inputProjectRoot),
    appMain: getEntryPoint(inputProjectRoot),
    modules: modulesPath,
    servedPath: getServedPath(inputProjectRoot),
    template: {
      get: templatePath,
      folder: templatePath(),
      indexHtml: templatePath('index.html'),
      manifest: templatePath('manifest.json'),
      serveJson: templatePath('serve.json'),
      favicon: templatePath('favicon.ico'),
      serviceWorker: templatePath('expo-service-worker.js'),
      registerServiceWorker: templatePath('register-service-worker.js'),
    },
    production: {
      get: getProductionPath,
      folder: getProductionPath(),
      indexHtml: getProductionPath('index.html'),
      manifest: getProductionPath('manifest.json'),
      serveJson: getProductionPath('serve.json'),
      favicon: getProductionPath('favicon.ico'),
      serviceWorker: getProductionPath('expo-service-worker.js'),
      registerServiceWorker: getProductionPath('register-service-worker.js'),
    },
  };
}

export function getEntryPoint(projectRoot: string): string | null {
  const { exp, pkg } = readConfigJson(projectRoot, true, true);

  /**
   *  The main file is resolved like so:
   * * `app.json` -> `expo.entryPoint`
   * * `package.json` -> `"main"`
   * * `possibleMainFiles`
   */
  if (exp && exp.entryPoint && typeof exp.entryPoint === 'string') {
    return getAbsolutePathWithProjectRoot(projectRoot, exp.entryPoint);
  } else if (pkg) {
    const { main } = pkg;
    if (main && typeof main === 'string') {
      return getAbsolutePathWithProjectRoot(projectRoot, main);
    }
    // Adds support for create-react-app (src/index.js) and react-native-cli (index.js) which don't define a main.
    for (const fileName of possibleMainFiles) {
      const filePath = getAbsolutePathWithProjectRoot(projectRoot, fileName);
      if (fs.existsSync(filePath)) {
        return filePath;
      }
    }
  }
  return null;
}

export function getPaths(projectRoot: string): FilePaths {
  const { exp } = readConfigJson(projectRoot, true, true);
  return parsePaths(projectRoot, exp);
}

export async function getPathsAsync(projectRoot: string): Promise<FilePaths> {
  let exp;
  try {
    exp = (await readConfigJsonAsync(projectRoot, true, true)).exp;
  } catch (error) {}
  return parsePaths(projectRoot, exp);
}

export function getServedPath(projectRoot: string): string {
  const { pkg } = readConfigJson(projectRoot, true, true);
  const envPublicUrl = process.env.WEB_PUBLIC_URL;

  // We use `WEB_PUBLIC_URL` environment variable or "homepage" field to infer
  // "public path" at which the app is served.
  // Webpack needs to know it to put the right <script> hrefs into HTML even in
  // single-page apps that may serve index.html for nested URLs like /todos/42.
  // We can't use a relative path in HTML because we don't want to load something
  // like /todos/42/static/js/bundle.7289d.js. We have to know the root.
  const publicUrl = envPublicUrl || pkg.homepage;
  const servedUrl = envPublicUrl || (publicUrl ? url.parse(publicUrl).pathname : '/');
  return ensureSlash(servedUrl!, true);
}

export function getPublicPaths({
  projectRoot,
  ...env
}: Environment): {
  /**
   * Webpack uses `publicPath` to determine where the app is being served from.
   * It requires a trailing slash, or the file assets will get an incorrect path.
   * In development, we always serve from the root. This makes config easier.
   */
  publicPath: string;

  /**
   * `publicUrl` is just like `publicPath`, but we will provide it to our app
   * as %WEB_PUBLIC_URL% in `index.html` and `process.env.WEB_PUBLIC_URL` in JavaScript.
   * Omit trailing slash as %WEB_PUBLIC_URL%/xyz looks better than %WEB_PUBLIC_URL%xyz.
   */
  publicUrl: string;
} {
  if (getMode(env) === 'production') {
    const publicPath = getServedPath(projectRoot);
    return {
      publicPath,
      publicUrl: publicPath.slice(0, -1),
    };
  }

  return { publicUrl: '', publicPath: '/' };
}

export function getProductionPath(projectRoot: string): string {
  const { exp } = readConfigJson(projectRoot, true, true);
  return getAbsolutePathWithProjectRoot(projectRoot, getWebOutputPath(exp));
}
