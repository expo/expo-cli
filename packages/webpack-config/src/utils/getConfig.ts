import { ExpoConfig, getConfigForPWA } from '@expo/config';

import { Environment } from '../types';
import { getPaths } from './paths';

function getConfig(env: Environment): ExpoConfig {
  if (env.config) {
    return env.config;
  }
  const locations = env.locations || getPaths(env.projectRoot);
  // Fill all config values with PWA defaults
  return getConfigForPWA(env.projectRoot, locations.absolute, {
    templateIcon: locations.template.get('icon.png'),
  });
}

export default getConfig;
