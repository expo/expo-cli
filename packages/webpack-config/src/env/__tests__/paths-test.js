/* eslint-env node */

import path from 'path';
import { getPaths, getPathsAsync, getPublicPaths, getServedPath } from '../paths';
import { normalizePaths } from '../../utils';

const projectRoot = path.resolve(__dirname, '../../../tests/basic');
const projectRootCustomHomepage = path.resolve(__dirname, '../../../tests/custom-homepage');
const projectRootMinimum = path.resolve(__dirname, '../../../tests/minimum');

function defaultNormalize(locations) {
  return normalizePaths(locations, value => value.split('packages/webpack-config/').pop());
}

beforeEach(() => {
  delete process.env.WEB_PUBLIC_URL;
});

it(`works with no app.json`, async () => {
  const locations = await getPathsAsync(projectRootMinimum);

  const normalized = defaultNormalize(locations);
  expect(normalized).toMatchSnapshot();

  expect(locations.servedPath).toBe('/');
});

it(`has consistent defaults`, async () => {
  const locations = await getPathsAsync(projectRoot);

  const normalized = defaultNormalize(locations);
  expect(normalized).toMatchSnapshot();

  expect(locations.servedPath).toBe('/');
});

it(`matches sync and async results`, async () => {
  const locations = getPaths(projectRoot);
  const normalized = defaultNormalize(locations);

  const locationsAsync = await getPathsAsync(projectRoot);
  const normalizedAsync = defaultNormalize(locationsAsync);

  expect(JSON.stringify(normalized)).toBe(JSON.stringify(normalizedAsync));
});

it(`uses a custom public path from WEB_PUBLIC_URL over the homepage field from package.json`, async () => {
  process.env.WEB_PUBLIC_URL = 'WEB_PUBLIC_URL-defined';
  const servedPath = await getServedPath(projectRootCustomHomepage);
  expect(servedPath).toBe('WEB_PUBLIC_URL-defined/');
});

it(`uses a custom public path from the homepage field of a project's package.json`, async () => {
  const servedPath = await getServedPath(projectRootCustomHomepage);
  expect(servedPath).toMatchSnapshot();
});

it(`changes public paths based on mode`, async () => {
  process.env.WEB_PUBLIC_URL = 'WEB_PUBLIC_URL-defined';

  const devPaths = await getPublicPaths({
    projectRoot: projectRootCustomHomepage,
    mode: 'development',
  });

  const prodPaths = await getPublicPaths({
    projectRoot: projectRootCustomHomepage,
    mode: 'production',
  });

  expect(devPaths.publicPath).not.toBe(prodPaths.publicPath);
  expect(devPaths.publicUrl).not.toBe(prodPaths.publicUrl);

  expect(prodPaths.publicUrl).toBe(process.env.WEB_PUBLIC_URL);
});

// TODO: Bacon: Add test for resolving entry point
// TODO: Bacon: Add test for custom config paths
