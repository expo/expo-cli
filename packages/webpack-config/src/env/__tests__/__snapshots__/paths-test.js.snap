// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`has consistent defaults 1`] = `
Object {
  "absolute": [Function],
  "appMain": "e2e/basic/index.js",
  "includeModule": [Function],
  "modules": "e2e/basic/node_modules",
  "packageJson": "e2e/basic/package.json",
  "production": Object {
    "favicon": "e2e/basic/web-build/favicon.ico",
    "folder": "e2e/basic/web-build",
    "get": [Function],
    "indexHtml": "e2e/basic/web-build/index.html",
    "manifest": "e2e/basic/web-build/manifest.json",
    "registerServiceWorker": "e2e/basic/web-build/register-service-worker.js",
    "serveJson": "e2e/basic/web-build/serve.json",
    "serviceWorker": "e2e/basic/web-build/expo-service-worker.js",
  },
  "root": "e2e/basic",
  "servedPath": "/",
  "template": Object {
    "favicon": "web-default/favicon.ico",
    "folder": "web-default",
    "get": [Function],
    "indexHtml": "web-default/index.html",
    "manifest": "web-default/manifest.json",
    "registerServiceWorker": "web-default/register-service-worker.js",
    "serveJson": "web-default/serve.json",
    "serviceWorker": "web-default/expo-service-worker.js",
  },
}
`;

exports[`uses a custom public path from the homepage field of a project's package.json 1`] = `"/homepage-defined/"`;

exports[`works with no app.json 1`] = `
Object {
  "absolute": [Function],
  "appMain": "e2e/minimum/index.js",
  "includeModule": [Function],
  "modules": "e2e/minimum/node_modules",
  "packageJson": "e2e/minimum/package.json",
  "production": Object {
    "favicon": "e2e/minimum/web-build/favicon.ico",
    "folder": "e2e/minimum/web-build",
    "get": [Function],
    "indexHtml": "e2e/minimum/web-build/index.html",
    "manifest": "e2e/minimum/web-build/manifest.json",
    "registerServiceWorker": "e2e/minimum/web-build/register-service-worker.js",
    "serveJson": "e2e/minimum/web-build/serve.json",
    "serviceWorker": "e2e/minimum/web-build/expo-service-worker.js",
  },
  "root": "e2e/minimum",
  "servedPath": "/",
  "template": Object {
    "favicon": "web-default/favicon.ico",
    "folder": "web-default",
    "get": [Function],
    "indexHtml": "web-default/index.html",
    "manifest": "web-default/manifest.json",
    "registerServiceWorker": "web-default/register-service-worker.js",
    "serveJson": "web-default/serve.json",
    "serviceWorker": "web-default/expo-service-worker.js",
  },
}
`;
